drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2023 EPAM Systems
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT regulator_gpio
+
+#include <stdint.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(regulator_gpio, CONFIG_REGULATOR_LOG_LEVEL);
+
+struct regulator_gpio_config {
+	struct regulator_common_config common;
+	struct gpio_dt_spec switch_volt;
+	int initial_state;
+	union {
+		const int *states_raw;
```

Comment:
I doubt that this will pass Zephyr review.  Zephyr tries to conform to MISRA and the way you use this `union` leads to both unspecified and undefined behaviour:

> - The value of a union member other than the last one stored into 

and 

> - if the value being stored in an object is read from another object that overlaps in any way
the storage of the first object, then the overlap shall be exact and the two objects shall
have qualified or unqualified versions of a compatible type; otherwise, the behavior is
undefined.

Better to leave previous implementation.

Reply:
I don't see problems here, both objects are pointers to some data so they have the same size and same overlapping